src/netbase.h

```diff
@@ -64,6 +65,63 @@ struct ProxyCredentials
     std::string password;
 };
 
+/**
+ * List of reachable networks. Everything is reachable by default.
+ */
+class ReachableNets {
+public:
+    void Add(Network net) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        AssertLockNotHeld(m_mutex);
+        if (net != NET_UNROUTABLE && net != NET_INTERNAL) {
+            LOCK(m_mutex);
+            m_reachable.insert(net);
+        }
+    }
+
+    void Remove(Network net) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        AssertLockNotHeld(m_mutex);
+        if (net != NET_UNROUTABLE && net != NET_INTERNAL) {
+            LOCK(m_mutex);
+            m_reachable.erase(net);
+        }
+    }
+
+    void RemoveAll() EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        LOCK(m_mutex);
+        m_reachable.clear();
```

Comment:
This will remove `NET_UNROUTABLE ` and `NET_INTERNAL`, after which they cannot be re-added. Is that intentional (the function name implies it is intentional)? I ask because the other accessors guard against their removal and insertion.

Reply:
Good question. Yes, it should be like this to preserve the behavior from `master` wrt which networks are reachable if `-onlynet` is used. The relevant snippet from this PR:

```diff
     if (args.IsArgSet("-onlynet")) {
-        std::set<enum Network> nets;
+        g_reachable_nets.RemoveAll();
         for (const std::string& snet : args.GetArgs("-onlynet")) {
             enum Network net = ParseNetwork(snet);
             if (net == NET_UNROUTABLE)
                 return InitError(strprintf(_("Unknown network specified in -onlynet: '%s'"), snet));
-            nets.insert(net);
-        }
-        for (int n = 0; n < NET_MAX; n++) {
-            enum Network net = (enum Network)n;
-            assert(IsReachable(net));
-            if (!nets.count(net))
-                SetReachable(net, false);
+            g_reachable_nets.Add(net);
         }
     }
```

By default all networks are reachable (in `master` and in this PR), including the dummy `NET_UNROUTABLE` and `NET_INTERNAL`. If `-onlynet=X` is used (where `X` can be `ipv4`, `ipv6`, `tor`, `i2p`, `cjdns`) then just `X` is made reachable and others unreachable (`NET_UNROUTABLE` and `NET_INTERNAL` also become unreachable).

As for the `Add()` and `Remove()` methods - I wanted to preserve the behavior of the old `SetReachable()`:

```diff
-void SetReachable(enum Network net, bool reachable)
-{
-    if (net == NET_UNROUTABLE || net == NET_INTERNAL)
-        return;
```

but no caller is ever calling `SetReachable()` with `NET_UNROUTABLE` or `NET_INTERNAL`, so I have removed those checks from the new `Add()` and `Remove()` methods.

Thanks!